### PR TITLE
SelectEventPoll: temporary restore to select efds

### PR DESCRIPTION
**Pull request by kwkam**
after 1fe6cc7 cpu usage was drastically increased

### CODE REVIEW

File: src/SelectEventPoll.cc [hunk 1]
```diff
@@ -174,6 +174,18 @@ void SelectEventPoll::poll(const struct timeval& tv)
 
   memcpy(&rfds, &rfdset_, sizeof(fd_set));
   memcpy(&wfds, &wfdset_, sizeof(fd_set));
+#ifdef __MINGW32__
+  fd_set efds;
+  FD_ZERO(&efds);
```
Comment (tatsuhiro-t) [inline src/SelectEventPoll.cc]: I'd like to know what happens if `memcpy(&efds, &wfdset_, sizeof(fd_set));` is inserted here.
Does it still cause high cpu usage?

**Comment by tatsuhiro-t** [comment]
Thank you for pointing this out. Could you check my comment?

**Comment by tatsuhiro-t** [comment]
How about this patch?

```patch
diff --git a/src/SelectEventPoll.cc b/src/SelectEventPoll.cc
index 824c5088..732a4596 100644
--- a/src/SelectEventPoll.cc
+++ b/src/SelectEventPoll.cc
@@ -174,6 +174,12 @@ void SelectEventPoll::poll(const struct timeval& tv)
 
   memcpy(&rfds, &rfdset_, sizeof(fd_set));
   memcpy(&wfds, &wfdset_, sizeof(fd_set));
+
+#ifdef __MINGW32__
+  fd_set efds;
+  memcpy(&efds, &wfdset_, sizeof(fd_set));
+#endif // __MINGW32__
+
 #ifdef ENABLE_ASYNC_DNS
 
   for (auto& i : nameResolverEntries_) {
@@ -190,11 +196,9 @@ void SelectEventPoll::poll(const struct timeval& tv)
   do {
     struct timeval ttv = tv;
 #ifdef __MINGW32__
-    // winsock will report non-blocking connect() errors in exceptfds, unlike
-    // posix, which will mark such sockets as writable.
-    // So just pass in our write socket set to exceptfds, too, to get connect()
-    // error notifications on Windows.
-    retval = select(fdmax_ + 1, &rfds, &wfds, &wfds, &ttv);
+    // winsock will report non-blocking connect() errors in efds,
+    // unlike posix, which will mark such sockets as writable.
+    retval = select(fdmax_ + 1, &rfds, &wfds, &efds, &ttv);
 #else // !__MINGW32__
     retval = select(fdmax_ + 1, &rfds, &wfds, nullptr, &ttv);
 #endif // !__MINGW32__
@@ -209,6 +213,11 @@ void SelectEventPoll::poll(const struct timeval& tv)
       if (FD_ISSET(e.getSocket(), &wfds)) {
         events |= EventPoll::EVENT_WRITE;
       }
+#ifdef __MINGW32__
+      if (FD_ISSET(e.getSocket(), &efds)) {
+        events |= EventPoll::EVENT_WRITE;
+      }
+#endif // __MINGW32__
       e.processEvents(events);
     }
   }
```

**Comment by kwkam** [comment]
Yes cpu usage is still high even `efds` is a different set.

**Comment by tatsuhiro-t** [comment]
Thanks.  Could you confirm that https://github.com/aria2/aria2/pull/1060#issuecomment-341968755 still causes high CPU usage?

I still could not reproduce this issue so far.  Could you share how to reproduce it?

**Comment by kwkam** [comment]
I can confirm that #1060 (comment) still cause high cpu.
This issue occurs when using BitTorrent. For example:
`aria2c https://images.offensive-security.com/kali-linux-2017.2-amd64.torrent`

**Comment by tatsuhiro-t** [comment]
Thank you.  I think the following patch fixes this issue:
```patch
diff --git a/src/SelectEventPoll.cc b/src/SelectEventPoll.cc
index 824c5088..5f6f3309 100644
--- a/src/SelectEventPoll.cc
+++ b/src/SelectEventPoll.cc
@@ -174,6 +174,12 @@ void SelectEventPoll::poll(const struct timeval& tv)
 
   memcpy(&rfds, &rfdset_, sizeof(fd_set));
   memcpy(&wfds, &wfdset_, sizeof(fd_set));
+
+#ifdef __MINGW32__
+  fd_set efds;
+  memcpy(&efds, &wfdset_, sizeof(fd_set));
+#endif // __MINGW32__
+
 #ifdef ENABLE_ASYNC_DNS
 
   for (auto& i : nameResolverEntries_) {
@@ -190,11 +196,9 @@ void SelectEventPoll::poll(const struct timeval& tv)
   do {
     struct timeval ttv = tv;
 #ifdef __MINGW32__
-    // winsock will report non-blocking connect() errors in exceptfds, unlike
-    // posix, which will mark such sockets as writable.
-    // So just pass in our write socket set to exceptfds, too, to get connect()
-    // error notifications on Windows.
-    retval = select(fdmax_ + 1, &rfds, &wfds, &wfds, &ttv);
+    // winsock will report non-blocking connect() errors in efds,
+    // unlike posix, which will mark such sockets as writable.
+    retval = select(fdmax_ + 1, &rfds, &wfds, &efds, &ttv);
 #else // !__MINGW32__
     retval = select(fdmax_ + 1, &rfds, &wfds, nullptr, &ttv);
 #endif // !__MINGW32__
@@ -209,6 +213,11 @@ void SelectEventPoll::poll(const struct timeval& tv)
       if (FD_ISSET(e.getSocket(), &wfds)) {
         events |= EventPoll::EVENT_WRITE;
       }
+#ifdef __MINGW32__
+      if (FD_ISSET(e.getSocket(), &efds)) {
+        events |= EventPoll::EVENT_ERROR;
+      }
+#endif // __MINGW32__
       e.processEvents(events);
     }
   }
```

**Comment by kwkam** [comment]
Yes https://github.com/aria2/aria2/pull/1060#issuecomment-341981822 fixed the issue.
Thank you very much.

**Comment by tatsuhiro-t** [comment]
Thank you for helping us.  I'll commit the patch.

**Pull request closed without merging**